web/src/narrow.js

```diff
@@ -1054,65 +1028,92 @@ export function deactivate(coming_from_recent_topics = false) {
+        $("#zfilt").removeClass("focused_table");
+        $("#zhome").addClass("focused_table");
+        message_lists.set_current(message_lists.home);
+        message_lists.current.resume_reading();
+        condense.condense_and_collapse($("#zhome div.message_row"));
+
+        reset_ui_state();
+        handle_middle_pane_transition();
+        hashchange.save_narrow();
+
+        if (message_lists.current.selected_id() !== -1) {
+            const preserve_pre_narrowing_screen_position =
+                message_lists.current.selected_row().length > 0 &&
+                message_lists.current.pre_narrow_offset !== undefined;
+            let message_id_to_select;
+            const select_opts = {
+                then_scroll: true,
+                use_closest: true,
+                empty_ok: true,
+            };
+
+            // We fall back to the closest selected id, if the user has removed a
+            // stream from the home view since leaving it the old selected id might
+            // no longer be there
+            // Additionally, we pass empty_ok as the user may have removed **all** streams
+            // from their home view
+            if (unread.messages_read_in_narrow) {
+                // We read some unread messages in a narrow. Instead of going back to
+                // where we were before the narrow, go to our first unread message (or
+                // the bottom of the feed, if there are no unread messages).
+                message_id_to_select = message_lists.current.first_unread_message_id();
+            } else {
+                // We narrowed, but only backwards in time (ie no unread were read). Try
+                // to go back to exactly where we were before narrowing.
+                if (preserve_pre_narrowing_screen_position) {
+                    // We scroll the user back to exactly the offset from the selected
+                    // message that they were at the time that they narrowed.
+                    // TODO: Make this correctly handle the case of resizing while narrowed.
+                    select_opts.target_scroll_offset = message_lists.current.pre_narrow_offset;
+                }
+                message_id_to_select = message_lists.current.selected_id();
             }
-            message_id_to_select = message_lists.current.selected_id();
+            message_lists.current.select_id(message_id_to_select, select_opts);
         }
-        message_lists.current.select_id(message_id_to_select, select_opts);
-    }
 
-    handle_post_narrow_deactivate_processes();
+        handle_post_narrow_deactivate_processes();
 
-    unnarrow_times.initial_core_time = new Date();
-    setTimeout(() => {
-        resize.resize_stream_filters_container();
-        unnarrow_times.initial_free_time = new Date();
-        report_unnarrow_time();
-    });
+        const free_span = span.startChild({
+            op: "function",
+            description: "unnarrow finish",
```

Comment:
Should this be described as "post unnarrow busy time", since I think that's what the time after this point is measuring (a paint/timers cycle)?

Reply:
Sure.  There's also a question of if we just want it around the `resize.resize_stream_filters_container()` call or if we want it across the `setTimeout` call.

Reply:
I'm not sure -- I'm not sure I have a clear idea of what `setTimeout(0` waits for, e.g. if it's possible for a paint to happen in the meantime, but my recollection is the aim was to measure separately (1) the time during which JavaScript is working and control is not returned to the browser -- that's the easiest to profile directly via the JS profiling tool in the browser if we see a regression, and (2) the rest of the time before the browser is idle, which will mostly be painting time, since I think we don't actually have a lot of potentially expensive callbacks triggered by narrowing.

Hmm, saying that out loud, I wonder if it'd be better to have the first time window end when this span starts, and plan to add the two numbers if we want the end-to-end total. I think that might be simpler in the code?

Reply:
I think finding the two spans to add together may be easier said than done, and certainly isn't something that one can run any aggregates or percentiles on.  The to transaction(s) are what we get aggregates on.  Right now, that includes the post-unnarrow busy time:
<img width="1057" alt="Screenshot 2023-05-04 at 8 29 34 PM" src="https://user-images.githubusercontent.com/28347/236356819-1dbf0321-4cff-4b36-b97b-fa9a6c5e9365.png">

Reply:
Well I think overall I'm content to change the infrastructure and come back to trying to optimize exactly how we do the accounting here when we have attention to focus on optimizing narrowing performance; I think even without the spans split up nicely we'll get plenty of information on whether certain classes of regressions happen, which we weren't getting in practice before.

Reply:
For reference, we can get the length of various sub-spans, including what percent of the time they exist:
<img width="1436" alt="Screenshot 2023-05-08 at 10 21 50 AM" src="https://user-images.githubusercontent.com/28347/236849565-2dd610c6-a6de-4ecf-a69b-cf572ae9231c.png">